dotnet/src/SemanticKernel/SkillDefinition/SKFunctionExtensions.cs

```diff
@@ -87,37 +86,51 @@ public static ISKFunction UseFrequencyPenalty(this ISKFunction skFunction, doubl
-        ContextVariables input,
-        ISemanticTextMemory memory,
-        IReadOnlySkillCollection? skills,
-        ILogger log,
-        CancellationToken cancellationToken = default)
+    /// <param name="input">Main input string</param>
+    /// /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>The result of the function execution</returns>
+    public static Task<SKContext> InvokeAsync(this ISKFunction function,
+        string? input = null, CancellationToken cancellationToken = default)
     {
-        var tmpContext = new SKContext(input, memory, skills, log, cancellationToken);
-        try
-        {
-#pragma warning disable CA2016 // the token is passed in via the context
-            await function.InvokeAsync(tmpContext).ConfigureAwait(false);
-#pragma warning restore CA2016
-        }
-        catch (Exception ex) when (!ex.IsCriticalException())
+        return function.InvokeAsync(input, settings: null, memory: null, logger: null, cancellationToken: cancellationToken);
+    }
+
+    /// <summary>
+    /// Execute a function allowing to pass the main input separately from the rest of the context.
+    /// Note: if the context contains an INPUT key/value, that value is ignored, logging a warning.
+    /// </summary>
+    /// <param name="function">Function to execute</param>
+    /// <param name="input">Main input string</param>
+    /// <param name="context">Execution context, including variables other than input</param>
+    /// <param name="settings">LLM completion settings (for semantic functions only)</param>
+    /// <returns>The result of the function execution</returns>
+    public static Task<SKContext> InvokeAsync(this ISKFunction function,
+        string input,
+        SKContext context,
+        CompleteRequestSettings? settings = null)
+    {
+        // Log a warning if the given input is overriding a different input in the context
+        if (context.Variables.ContainsKey(ContextVariables.MainKey) && !string.IsNullOrEmpty(context.Variables.Input))
         {
-            log.LogError(ex, "Something went wrong when invoking function with custom input: {0}.{1}. Error: {2}", function.SkillName,
-                function.Name, ex.Message);
-            tmpContext.Fail(ex.Message, ex);
+            if (string.Equals(input, context.Variables.Input, StringComparison.Ordinal))
+            {
+                context.Log.LogWarning(
+                    "Function {0}.{1} has been invoked with an explicit input text that is different and overrides the input text defined in the context",
+                    function.SkillName, function.Name);
+            }
         }
 
-        return tmpContext;
+        // Create a copy of the context, to avoid editing the original set of variables
+        SKContext contextClone = context.Clone();
```

Comment:
Passing a clone of SKContext into the function: this differs from the Invoke* methods in SKFunction and Plan, where the actual context object is passed into the function and updated (mutable). Here, it's always a shallow clone (partially immutable), which introduces some different behavior/expectation of SKContext usage.

Reply:
This is in line with the other calls, happy to chat about this offline. 

When a function receives in input a context and returns a context, the 2 are not the same. Changing variables in the input-context is different from setting variables in the output-context. In fact, we should make the input-context read-only to avoid similar attempts.

Reply:
pls see `ContextSwitchInSKContextOutTaskSKContext`

```
context = await callable(context).ConfigureAwait(false);
return context;
```

the context returned is not the mutable context sent in, but the context returned by `callable`

Reply:
I added a parameter to allow both behaviors, so one can run the function with immutable state, e.g. to compare results. The default behavior is allowing functions to change it (mutable context)